File: d15/d15_2.py
# input = [0, 3, 6]
input = [0, 13, 16, 17, 1, 10, 6]

spoken = {}

turn = 1
for num in input:
    spoken[num] = [turn, -1]
    turn += 1
lastNum = input[-1]


def updateSpoken(turn, num):
    if num in spoken:
        oldInfo = spoken[num]
        spoken[num] = [turn, oldInfo[0]]
    else:
        spoken[num] = [turn, -1]


def wasSpoken(turn, num):
    info = spoken[num]
    return info and ((info[0] != turn - 1) or (info[1] != -1))


while True:
    if wasSpoken(turn, lastNum):
        info = spoken[lastNum]
        spokenNum = (turn - 1) - info[1]
        firstTime = True
    else:
        firstTime = False
        spokenNum = 0
    updateSpoken(turn, spokenNum)
    turn += 1
    lastNum = spokenNum

    if turn == 30000000 + 1:
        break

print("result={}".format(spokenNum))

File: d15/d15_2.js
var input = [0, 3, 6];
//var input = [0, 13, 16, 17, 1, 10, 6];

var spoken = {};
var turn;

function Setup() {
	turn = 1;
	for (var t = 0; t < input.length; t++) {
		spoken[input[t]] = [turn, -1];
		turn++;
	}
	lastNum = input[input.length - 1];
}

function updateSpoken(turn, num) {
	var oldInfo = spoken[num];
	if (oldInfo) {
		spoken[num] = [turn, oldInfo[0]];
	} else {
		spoken[num] = [turn, -1];
	}
}

function wasSpoken(turn, num) {
	var info = spoken[num];
	return info && ((info[0] != turn - 1) || (info[1] != -1));
}

// !!!!!!!!!!!!!!!!!!!!!!!!!!!!
// !!! JS ran out of memory !!!
// !!!!!!!!!!!!!!!!!!!!!!!!!!!!
function Loop() {
	var spokenNum;
	var firstTime;

	while (true) {
		if (wasSpoken(turn, lastNum)) {
			var info = spoken[lastNum];
			spokenNum = (turn - 1) - info[1];
			firstTime = true;
		} else {
			firstTime = false;
			spokenNum = 0;
		}
		updateSpoken(turn, spokenNum);

		turn++;
		lastNum = spokenNum;

		if ((turn % 10000) == 0) {
			Println(turn)
		}

		if (turn == 30000000 + 1) { // we are interested in the 30000000th number
			break;
		}
	}

	Println("result=" + spokenNum);

	Stop();
}

function Input(e) { }

